apps/src/code-studio/initApp/project.js

```diff
@@ -1923,6 +1923,23 @@ var projects = (module.exports = {
     }
     return sourcesApi;
   },
+
+  // Sets a callback to save the project before unloading the page.
+  registerSaveOnUnload() {
+    window.addEventListener('beforeunload', this.unloadHandler_.bind(this));
```

Comment:
We do not...is there a specific problem you're concerned about / alternative approach you'd suggest?

We add the listener once on mount, and basically want to keep it there until you navigate away from the page, so I'm not sure when we'd remove it?

Reply:
Tested locally, looks like my concerns do not confirm since there's no way to naigate away from app lab screen rather then reloading the page fully (native web page chenge vs react no reload page/screens change).

I was worried about that this eventListener might be triggered on pages you visit after the app lab page, but looks like it's not possible. Also, if project is not changed, autosave on close is not triggered, which is awesome! In docs, removeEventListener was involved in that logic, but in our case we've did that without removeEventListener so we don't need one in the current code.

Though, reading more through docs - if we do want to add this functionality for mobile/tablets also - looks like it might be worth using `visibilitychange` instead of `beforeunload`. Since:

```
The beforeunload event suffers from some problems:

It is not reliably fired, especially on mobile platforms. For example, the beforeunload event is not fired at all in the following scenario:
A mobile user visits your page.
The user then switches to a different app.
Later, the user closes the browser from the app manager.
Note: It is recommended to use the [visibilitychange](https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event)
 event as a more reliable signal for automatic app state saving that gets around problems like the above.
 See [Don't lose user and app state, use Page Visibility](https://www.igvita.com/2015/11/20/dont-lose-user-and-app-state-use-page-visibility/) for more details.
```


https://developer.mozilla.org/en-US/docs/Web/API/Document/visibilitychange_event

What do you think?

Reply:
I'm not super worried about it (this is only relevant for a maximum of 30 seconds of work, since we autosave on that interval), the mobile/tablet case sounds pretty niche. Good to know re: visibilitychange though for future reference!